examples-da/partial/importance_weighted_adversarial_nets.py

```diff
@@ -0,0 +1,415 @@
+    D_0 = DomainDiscriminator(in_feature=classifier_s.features_dim, hidden_size=1024).to(device)
+
+    # stage 2: train feature_extractor F_t and domain classifier D, D_0 simultaneously
+    train(train_source_iter, train_target_iter, classifier_s, classifier_t, D, D_0, val_loader, test_loader, args)
+
+
+def pretrain_classifier(model: ImageClassifier, dataset, args: argparse.Namespace):
+    """In this function we perform supervised pretraining
+    Inputs:
+        **model**: model to pretrain(classifier_s)
+        **dataset**: data from source domain
+    """
+    num_samples = len(dataset)
+    num_samples_train = num_samples * 4 // 5
+    # split dataset into training_set and validation_set
+    train_dataset, val_dataset = torch.utils.data.random_split(dataset,
+                                                               [num_samples_train, num_samples - num_samples_train])
+
+    train_loader = DataLoader(train_dataset, batch_size=36, shuffle=True, num_workers=4, drop_last=True)
+    val_loader = DataLoader(val_dataset, batch_size=36, shuffle=False, num_workers=4)
+
+    train_iter = ForeverDataIterator(train_loader)
+
+    # parameters for pretraining
+    optimizer = SGD(model.get_parameters(), args.lr, momentum=args.momentum, weight_decay=args.weight_decay,
+                    nesterov=True)
+    lr_scheduler = LambdaLR(optimizer, lambda x: args.lr * (1. + args.lr_gamma * float(x)) ** (-args.lr_decay))
+
+    # start training
+    print('pretrain feature extractor F_s and classifier C on source domain.')
+    best_acc1 = 0.
+    for epoch in range(args.epochs):
+        # statistics
+        losses = AverageMeter('Loss', ':6.2f')
+        cls_accs = AverageMeter('Cls Acc', ':3.1f')
+        progress = ProgressMeter(
+            args.iters_per_epoch,
+            [losses, cls_accs],
+            prefix="Epoch: [{}]".format(epoch))
+        # switch to train mode
+        model.train()
+
+        for i in range(args.iters_per_epoch):
+            x, labels = next(train_iter)
+            x = x.to(device)
+            labels = labels.to(device)
+            
+            y, _ = model(x)
+
+            loss = F.cross_entropy(y, labels)
+            cls_acc = accuracy(y, labels)[0]
+            losses.update(loss.item(), x.size(0))
+            cls_accs.update(cls_acc.item(), x.size(0))
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            lr_scheduler.step()
+
+            if i % args.print_freq == 0:
+                progress.display(i)
+
+        # evaluate on validation set
+        acc1 = validate(val_loader, model, args)
+
+        # remember best acc@1 and save checkpoint
+        if acc1 > best_acc1:
+            best_model = copy.deepcopy(model.state_dict())
+        best_acc1 = max(acc1, best_acc1)
+
+    print("best_acc1 = {:3.1f}".format(best_acc1))
+    model.load_state_dict(best_model)
```

Comment:
It's not very fair to load best model as pretrain model since we don't have labels on the target domain.